File: test_dates.mjs
// Date-clamping tests for triage.html's inline script.
//
// The 2026-07-28 bug was "derive a calendar day in UTC", and it existed in both
// languages. test_dates.py guards the Python side, but it can only grep the
// browser's copy — it can't run it. This does.
//
// The specific regression this locks down: building today's date by slicing an
// ISO string returns the UTC day, so after 5pm Pacific the dashboard showed
// tomorrow's date on job cards and stamped tomorrow onto export filenames.
//
// Run: node test_dates.mjs      (also runs in tests.yml)

// localToday() reads the *device's* timezone — correct for a browser, since the
// viewer's "today" is the one that matters. But it means these assertions are
// only meaningful from a Pacific clock, and CI runners are UTC: without this
// line every clamp test below passes locally and fails in Actions. Set before
// any Date is constructed.
process.env.TZ = 'America/Los_Angeles';

import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';

const here = dirname(fileURLToPath(import.meta.url));
// NB: this file lives at the repo root, unlike sync/merge.test.mjs.
const html = readFileSync(join(here, 'triage.html'), 'utf8');

// Copied from sync/merge.test.mjs — that file exports nothing, runs its
// assertions at import time and calls process.exit(), so it can't be imported.
// Pull a top-level `function name(...) { ... }` out of the inline <script> by
// brace-matching from its opening brace. Regex alone can't do this safely.
function extractFunction(src, name) {
  const start = src.indexOf(`function ${name}(`);
  if (start === -1) throw new Error(`triage.html no longer defines ${name}() — did it get renamed?`);
  const open = src.indexOf('{', start);
  let depth = 0;
  for (let i = open; i < src.length; i++) {
    if (src[i] === '{') depth++;
    else if (src[i] === '}' && --depth === 0) return src.slice(start, i + 1);
  }
  throw new Error(`unbalanced braces while extracting ${name}()`);
}

const NAMES = ['localToday', 'displayDate'];
const client = new Function(
  `${NAMES.map(n => extractFunction(html, n)).join('\n')}
   return { ${NAMES.join(', ')} };`
)();

// 2026-07-29T01:30:00Z === 18:30 PDT on 2026-07-28 — inside the window where
// the UTC day and the Pacific day disagree. Fixed clock: these must not depend
// on wall time (same convention as sync/merge.test.mjs).
const EVENING_PT = Date.parse('2026-07-29T01:30:00Z');
const MORNING_PT = Date.parse('2026-07-28T17:00:00Z');  // 10:00 PDT, days agree

let failed = 0;
function check(name, actual, expected) {
  const ok = actual === expected;
  if (!ok) failed++;
  console.log(`${ok ? '  ok  ' : '  FAIL'} ${name}${ok ? '' : ` — got ${JSON.stringify(actual)}, want ${JSON.stringify(expected)}`}`);
}

// The headline case. Slicing toISOString() here would give '2026-07-29'.
check('localToday() at 6:30pm PDT returns the Pacific day',
  client.localToday(EVENING_PT), '2026-07-28');
check('localToday() mid-morning returns the same day',
  client.localToday(MORNING_PT), '2026-07-28');
check('localToday() zero-pads single-digit months and days',
  client.localToday(Date.parse('2026-03-05T20:00:00Z')), '2026-03-05');

// A date that hasn't happened yet is never correct — clamp it. This is what
// covers the ~34 rows already committed to all_jobs.json before the fix, which
// stay in the 14-day window.
check('a future date_posted is clamped to today',
  client.displayDate('2026-07-29', EVENING_PT), '2026-07-28');
check('a far-future date_posted is clamped to today',
  client.displayDate('2027-01-01', EVENING_PT), '2026-07-28');
check('a past date_posted is untouched',
  client.displayDate('2026-07-20', EVENING_PT), '2026-07-20');
check("today's date_posted is untouched",
  client.displayDate('2026-07-28', EVENING_PT), '2026-07-28');

// Workday emits these; rewriting them would destroy information.
check('relative strings pass through',
  client.displayDate('Posted Today', EVENING_PT), 'Posted Today');
check('relative day-count strings pass through',
  client.displayDate('Posted 9 Days Ago', EVENING_PT), 'Posted 9 Days Ago');
check('empty date_posted stays empty',
  client.displayDate('', EVENING_PT), '');
check('missing date_posted stays empty',
  client.displayDate(undefined, EVENING_PT), '');

// Recurrence guard, mirroring RecurrenceGuards in test_dates.py: the browser
// must never derive a calendar day from an ISO string slice.
if (/toISOString\(\)\s*\.\s*slice\(\s*0\s*,\s*10\s*\)/.test(html)) {
  console.log('  FAIL triage.html derives a day from toISOString().slice(0,10) — that is the UTC day');
  failed++;
} else {
  console.log('  ok   triage.html never slices toISOString() for a calendar day');
}

console.log(failed ? `\n${failed} FAILURE(S) in triage.html date handling`
                   : '\nAll triage.html date checks passed');
process.exit(failed ? 1 : 0);

File: test_dates.py
#!/usr/bin/env python3
"""
Tests for posted-date normalization in scrape_jobs.py.

The bug (2026-07-28): the watchers run on UTC GitHub Actions runners, so every
date_posted was a UTC calendar day. After 17:00 PDT that day is already
tomorrow locally, and 34 roles on the live dashboard were dated 2026-07-29 on
the evening of the 28th.

The headline test is test_choke_point_clamps_a_future_dated_job: every scraper
writes through save_jobs_output(), so normalizing there is what makes a source
we haven't written yet safe by default. Run it with:

    python test_dates.py

No network, no secrets, no dependencies (scrape_jobs' module-level imports are
all stdlib — jobspy is imported lazily inside the scrape functions).
"""

import json
import os
import sys
import tempfile
import unittest
from datetime import date, datetime, timedelta, timezone

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, SCRIPT_DIR)

import scrape_jobs as sj  # noqa: E402

TODAY = date(2026, 7, 28)


class NormalizePostedDate(unittest.TestCase):
    def norm(self, value):
        return sj.normalize_posted_date(value, today=TODAY)

    def test_utc_timestamp_becomes_the_pacific_day(self):
        # The whole bug in one assertion: 01:00Z on the 29th is 18:00 PDT on
        # the 28th, and the dashboard must say the 28th.
        self.assertEqual(self.norm("2026-07-29T01:00:00Z"), "2026-07-28")

    def test_offset_timestamp_converts(self):
        self.assertEqual(self.norm("2026-07-29T04:00:00+04:00"), "2026-07-28")

    def test_naive_timestamp_is_assumed_utc(self):
        # Must agree with _parse_posted_at, which also assumes UTC for naive
        # values. Two different answers for one string is how the writer and
        # the parser drift apart.
        self.assertEqual(self.norm("2026-07-29T01:00:00"), "2026-07-28")

    def test_bare_future_date_is_clamped(self):
        self.assertEqual(self.norm("2026-07-29"), "2026-07-28")

    def test_bare_past_date_is_untouched(self):
        self.assertEqual(self.norm("2026-07-20"), "2026-07-20")

    def test_relative_strings_pass_through(self):
        # The dashboard's jobDateMs() understands these; rewriting them would
        # destroy information.
        for s in ("Posted Today", "Posted 9 Days Ago", "Posted 30+ Days Ago"):
            self.assertEqual(self.norm(s), s)

    def test_empty_and_non_string_inputs(self):
        for v in ("", None, "   "):
            self.assertEqual(self.norm(v), "")

    def test_unparseable_input_survives(self):
        self.assertEqual(self.norm("sometime last week"), "sometime last week")

    def test_never_returns_a_future_day(self):
        for v in ("2026-12-25", "2027-01-01T00:00:00Z", "2026-07-29"):
            self.assertLessEqual(self.norm(v), TODAY.strftime("%Y-%m-%d"))

    def test_idempotent(self):
        # save_jobs_output normalizes on top of the probe-level fixes, so every
        # value gets normalized at least twice.
        for v in ("2026-07-29", "2026-07-29T01:00:00Z", "Posted Today", ""):
            once = self.norm(v)
            self.assertEqual(self.norm(once), once, v)


class FreshnessFilterInteraction(unittest.TestCase):
    """
    Regression guard for the trap this change walked into.

    Converting probe dates to Pacific shifts a bare date back by up to a day.
    _parse_posted_at used to read bare dates as UTC midnight, so a 30-minute-old
    role read as 25.5h old and is_recent_posting() dropped it against the 24h
    FRESH_JOB_LOOKBACK. Parsing bare dates as LOCAL_TZ midnight is what keeps
    the writer and the reader agreeing.
    """

    NOW = datetime(2026, 7, 29, 1, 30, tzinfo=timezone.utc)  # 18:30 PDT on the 28th

    def test_todays_pacific_date_is_still_fresh(self):
        self.assertTrue(sj.is_recent_posting({"date_posted": "2026-07-28"}, now=self.NOW))

    def test_bare_date_parses_as_local_midnight(self):
        parsed = sj._parse_posted_at("2026-07-28", now=self.NOW)
        self.assertEqual(parsed.utcoffset(), timedelta(hours=-7))
        self.assertLess(self.NOW - parsed, sj.FRESH_JOB_LOOKBACK)

    def test_naive_timestamp_still_parses_as_utc(self):
        # The other branch of the same if/else must NOT move to Pacific — a
        # naive timestamp is an instant, not a calendar day.
        parsed = sj._parse_posted_at("2026-07-29T01:00:00", now=self.NOW)
        self.assertEqual(parsed.utcoffset(), timedelta(0))

    def test_genuinely_stale_roles_are_still_dropped(self):
        self.assertFalse(sj.is_recent_posting({"date_posted": "2026-07-20"}, now=self.NOW))


class MaxAgePolicy(unittest.TestCase):
    """
    The 14-day ceiling added 2026-08-19: the ATS registry shipped with no
    date filter and surfaced 2024 reqs. These tests pin the two parser
    extensions that make the ceiling enforceable (Workday day-relative
    strings; Lever epoch-ms) and the keep-when-unprovable rule.
    """

    NOW = datetime(2026, 8, 19, 12, 0, tzinfo=timezone.utc)

    def stale(self, value):
        return sj.is_stale_posting(value, now=self.NOW)

    def test_workday_day_relative_strings_parse(self):
        self.assertEqual(
            sj._parse_posted_at("Posted 16 Days Ago", now=self.NOW),
            self.NOW - timedelta(days=16),
        )

    def test_workday_thirty_plus_floor_is_read_as_thirty(self):
        self.assertEqual(
            sj._parse_posted_at("Posted 30+ Days Ago", now=self.NOW),
            self.NOW - timedelta(days=30),
        )

    def test_weeks_and_months_parse(self):
        self.assertEqual(
            sj._parse_posted_at("Posted 2 Weeks Ago", now=self.NOW),
            self.NOW - timedelta(weeks=2),
        )
        self.assertEqual(
            sj._parse_posted_at("Posted 3 Months Ago", now=self.NOW),
            self.NOW - timedelta(days=90),
        )

    def test_lever_epoch_milliseconds_parse(self):
        expected = self.NOW - timedelta(days=2)
        ms = int(expected.timestamp() * 1000)
        self.assertEqual(sj._parse_posted_at(str(ms), now=self.NOW), expected)
        # Registry rows committed before 2026-08-19 carry the raw int.
        self.assertEqual(sj._parse_posted_at(ms, now=self.NOW), expected)

    def test_small_integers_are_not_dates(self):
        self.assertIsNone(sj._parse_posted_at("12345", now=self.NOW))

    def test_stale_boundary(self):
        self.assertFalse(self.stale("Posted 13 Days Ago"))
        self.assertTrue(self.stale("Posted 15 Days Ago"))
        self.assertTrue(self.stale("Posted 30+ Days Ago"))
        self.assertTrue(self.stale("2024-09-04"))

    def test_unprovable_staleness_is_kept(self):
        for value in ("", None, "next Tuesday"):
            self.assertFalse(self.stale(value), repr(value))

    def test_minutes_and_hours_still_parse(self):
        # The regex extension must not disturb the original relative branch.
        self.assertEqual(
            sj._parse_posted_at("Posted 3 hours ago", now=self.NOW),
            self.NOW - timedelta(hours=3),
        )


class ChokePoint(unittest.TestCase):
    """
    The important one. Every saver routes through save_jobs_output(), so this
    is what makes a source nobody has written yet safe by default.
    """

    def setUp(self):
        # save_jobs_output builds every output path from the module-level
        # SCRIPT_DIR, so os.chdir() does NOT redirect it. Without this patch the
        # test would overwrite the real linkedin_jobs.json / all_jobs.json.
        self.tmp = tempfile.mkdtemp()
        self.real_script_dir = sj.SCRIPT_DIR
        sj.SCRIPT_DIR = self.tmp
        # Don't page a developer who has Pushover configured locally.
        self.saved_env = {k: os.environ.pop(k, None)
                          for k in ("PUSHOVER_TOKEN", "PUSHOVER_USER")}

    def tearDown(self):
        sj.SCRIPT_DIR = self.real_script_dir
        for k, v in self.saved_env.items():
            if v is not None:
                os.environ[k] = v

    def save(self, jobs, basename="test_choke_point"):
        sj.save_jobs_output(
            jobs, basename=basename, title="t", subtitle="s",
            accent="#000", empty_message="none", window_label="w",
        )
        with open(os.path.join(self.tmp, f"{basename}.json")) as f:
            return json.load(f)["jobs"]

    def test_choke_point_clamps_a_future_dated_job(self):
        today = sj.local_today().strftime("%Y-%m-%d")
        future = (sj.local_today() + timedelta(days=1)).strftime("%Y-%m-%d")
        out = self.save([{
            "company": "Acme", "title": "Account Manager", "location": "Long Beach, CA",
            "url": "https://example.com/jobs/1", "date_posted": future,
            "ats": "LinkedIn",
        }])
        self.assertEqual(out[0]["date_posted"], today)

    def test_no_written_job_is_ever_future_dated(self):
        today = sj.local_today().strftime("%Y-%m-%d")
        jobs = [
            {"company": "A", "title": "Account Manager", "url": "https://example.com/1",
             "date_posted": "2027-01-01", "location": "Long Beach, CA", "ats": "LinkedIn"},
            {"company": "B", "title": "Account Manager", "url": "https://example.com/2",
             "date_posted": "2026-07-01", "location": "Long Beach, CA", "ats": "Indeed"},
            {"company": "C", "title": "Account Manager", "url": "https://example.com/3",
             "date_posted": "Posted Today", "location": "Long Beach, CA", "ats": "Workday"},
        ]
        for j in self.save(jobs):
            d = j.get("date_posted", "")
            if d and d[:4].isdigit():
                self.assertLessEqual(d, today, j["url"])

    def test_all_jobs_master_is_normalized_too(self):
        # The dashboard's Rank tab reads all_jobs.json, so the accumulator has
        # to see normalized values, not raw ones.
        future = (sj.local_today() + timedelta(days=1)).strftime("%Y-%m-%d")
        self.save([{
            "company": "Acme", "title": "Account Manager", "location": "Long Beach, CA",
            "url": "https://example.com/jobs/master", "date_posted": future,
            "ats": "LinkedIn",
        }])
        with open(os.path.join(self.tmp, "all_jobs.json")) as f:
            master = json.load(f)["jobs"]
        self.assertTrue(master)
        self.assertEqual(master[0]["date_posted"],
                         sj.local_today().strftime("%Y-%m-%d"))


class RecurrenceGuards(unittest.TestCase):
    """
    The bug wasn't one line — it was a habit of deriving calendar days in UTC.
    Fail the build if either idiom comes back, whether by revert or by someone
    copy-pasting an existing probe to add a new ATS.
    """

    FORBIDDEN = [
        ('or "")[:10]',
         'truncating an ISO timestamp yields a UTC day — use normalize_posted_date()'),
        ('tz=timezone.utc).strftime("%Y-%m-%d")',
         'derive calendar days in LOCAL_TZ, not UTC'),
    ]

    def test_scrape_jobs_has_no_utc_day_derivation(self):
        with open(os.path.join(SCRIPT_DIR, "scrape_jobs.py")) as f:
            body = f.read()
        for literal, why in self.FORBIDDEN:
            self.assertNotIn(literal, body, why)

    def test_utc_labelled_timestamps_are_untouched(self):
        # Guard the guard: these are explicitly labelled UTC and are correct.
        # If a future tightening of FORBIDDEN starts matching them, this fails.
        with open(os.path.join(SCRIPT_DIR, "scrape_jobs.py")) as f:
            body = f.read()
        self.assertIn('strftime("%Y-%m-%d %H:%M UTC")', body)
        for literal, _ in self.FORBIDDEN:
            self.assertNotIn(literal, 'strftime("%Y-%m-%d %H:%M UTC")')

    def test_triage_html_does_not_build_a_day_from_toisostring(self):
        # The same bug in the browser: toISOString() is the UTC day.
        with open(os.path.join(SCRIPT_DIR, "triage.html")) as f:
            body = f.read()
        self.assertNotIn("toISOString().slice(0, 10)", body)
        self.assertNotIn("toISOString().slice(0,10)", body)


if __name__ == "__main__":
    unittest.main(verbosity=2)
